xiilib/_gunicorn/wsgi_app.py

```diff
@@ -0,0 +1,131 @@
+            charm_state: The state of the charm.
+            webserver: The webserver manager object.
+            database_migration: The database migration manager object.
+        """
+        self._charm_state = charm_state
+        self._container = container
+        self._webserver = webserver
+        self._database_migration = database_migration
+
+    def _encode_env(self, value: str | int | float | bool | list | dict) -> str:
+        """Encode the environment variable values.
+
+        Args:
+            value: the input environment variable value.
+
+        Return:
+            The original string if the input is a string, or JSON encoded value.
+        """
+        return value if isinstance(value, str) else json.dumps(value)
+
+    def _wsgi_environment(self) -> dict[str, str]:
+        """Generate a WSGI environment dictionary from the charm WSGI configurations.
+
+        The WSGI environment generation follows these rules:
+            1. User-defined configuration cannot overwrite built-in WSGI configurations, even if
+                the built-in Flask configuration value is None (undefined).
+            2. Boolean and integer-typed configuration values will be JSON encoded before
+                being passed to application.
+            3. String-typed configuration values will be passed to the application as environment
+                variables directly.
+
+        Returns:
+            A dictionary representing the Flask environment variables.
+        """
+        config = self._charm_state.app_config
+        config.update(self._charm_state.wsgi_config)
+        prefix = f"{self._charm_state.framework.upper()}_"
+        env = {f"{prefix}{k.upper()}": self._encode_env(v) for k, v in config.items()}
+        secret_key_env = f"{prefix}SECRET_KEY"
+        if secret_key_env not in env:
+            env[secret_key_env] = self._charm_state.secret_key
+        for proxy_variable in ("http_proxy", "https_proxy", "no_proxy"):
+            proxy_value = getattr(self._charm_state.proxy, proxy_variable)
+            if proxy_value:
+                env[proxy_variable] = str(proxy_value)
+                env[proxy_variable.upper()] = str(proxy_value)
+        env.update(self._charm_state.database_uris)
+        return env
+
+    def _wsgi_layer(self) -> ops.pebble.LayerDict:
+        """Generate the pebble layer definition for WSGI application.
+
+        Returns:
+            The pebble layer definition for WSGI application.
+        """
+        original_services_file = self._charm_state.state_dir / "original-services.json"
+        if self._container.exists(original_services_file):
+            services = json.loads(self._container.pull(original_services_file).read())
+        else:
+            plan = self._container.get_plan()
+            services = {k: v.to_dict() for k, v in plan.services.items()}
+            self._container.push(original_services_file, json.dumps(services), make_dirs=True)
```

Comment:
App container, as the app container can be changed by updating the resources, and when app container changed, the saved file will go away with it which makes saved services consistent.

Reply:
And what about storing it in the peer relation data ? I'm not a big fan of keeping in it on the container FS

Reply:
I was thinking about this a bit more after Weii's reply and I think it is possible to, for example, just refresh the app container. If it is stored on the app container, then the original services file is correctly cleared since the original service might not be valid anymore. However, if it is stored in the peer relation or on the charm container, it would not be cleaned up. So storing it on the app container links the lifecycle of the file to the lifecycle of the original service definition in the rock file

Reply:
That makes sense indeed